pkg/controller/certificate-shim/gateways/controller.go

```diff
@@ -0,0 +1,148 @@
+/*
+Copyright 2020 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	gatewayapi "sigs.k8s.io/gateway-api/apis/v1alpha1"
+	gatewayclient "sigs.k8s.io/gateway-api/pkg/client/clientset/versioned"
+	gatewayinformers "sigs.k8s.io/gateway-api/pkg/client/informers/externalversions"
+	gapilisters "sigs.k8s.io/gateway-api/pkg/client/listers/apis/v1alpha1"
+
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	controllerpkg "github.com/jetstack/cert-manager/pkg/controller"
+	shimhelper "github.com/jetstack/cert-manager/pkg/controller/certificate-shim"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+)
+
+const (
+	ControllerName = "gateway-shim"
+
+	// resyncPeriod is set to 10 hours across cert-manager. These 10 hours come
+	// from a discussion on the controller-runtime project that boils down to:
+	// never change this without an explicit reason.
+	// https://github.com/kubernetes-sigs/controller-runtime/pull/88#issuecomment-408500629
+	resyncPeriod = 10 * time.Hour
+)
+
+type defaults struct {
+	autoCertificateAnnotations          []string
+	issuerName, issuerKind, issuerGroup string
+}
+
+type controller struct {
+	gatewayList gapilisters.GatewayLister
+	sync        func(context.Context, metav1.Object) error
+}
+
+func (c *controller) Register(ctx *controllerpkg.Context) (workqueue.RateLimitingInterface, []cache.InformerSynced, error) {
+	// The user may have enabled the gateway-shim controller but forgot to
+	// install the Gateway API CRDs. This will cause cert-manager to go into
+	// CrashLoopBackoff which is nice and obvious.
+	d, err := discovery.NewDiscoveryClientForConfig(ctx.RESTConfig)
+	if err != nil {
+		return nil, nil, fmt.Errorf("%s: couldn't construct discovery client: %w", ControllerName, err)
+	}
+	resources, err := d.ServerResourcesForGroupVersion(gatewayapi.GroupVersion.String())
+	if err != nil {
+		return nil, nil, fmt.Errorf("%s: couldn't discover gateway API resources (are the Gateway API CRDS installed?): %w", ControllerName, err)
```

Comment:
I changed the behavior to what you suggested.

In case the CRDs aren't found on startup, we let the user know that cert-manager will keep retrying looking for the CRDs with this message (shown a single time):
  
```
controller.go:181] cert-manager/controller/build-context "msg"="the
Gateway API CRDs do not seem to be present, cert-manager will keep
retrying watching for them"
```

The user then sees the following message printed (using an exponential back-off):

```
reflector.go:167: Failed to watch *v1alpha1.Gateway: failed to list
*v1alpha1.Gateway: the server could not find the requested resource
(get gateways.networking.x-k8s.io)
```